### PR TITLE
bcda-4731 Feature: Implement get jobs endpoint

### PR DESCRIPTION
**Pull request by SMLuthi**
### Fixes [BCDA-4731](https://jira.cms.gov/browse/BCDA-4731)

Currently, if an ACO runs a job and does not remember the job id they have no way of retrieving that information. They have to generate a new job adding extra load on the system.

### Proposed Changes

Create a new `GET jobs` endpoint that will retrieve historical data on the jobs that have been run by that ACO including the job id, time the job was created, time the job was completed, and the original request the job was for. The response should be in the fhir `Task` format.

### Change Details

- Create routing for `GET .../v1/jobs` and `GET .../v2/jobs`
- Allow route to take optional `_status` parameter that can take a comma separated set of values that pertain to our jobs statuses (ie `Completed`, `Failure`, etc...). Not providing a status endpoint will retrieve ALL job statuses.
- Example request `https://{{host}}api/v1/jobs?_status=Completed,Failed,Expired`
- Create new fhir Bundle writer that generates a fhir `Task` response when given the job(s) data.

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [x] new data stored or transmitted
We are sharing job information that the ACO has run in the past. We aren't sharing anything that hasn't already been shared with the ACO at some point (ie. when creating the job).

Example payload:
```
{
    "entry":[
        {
            "resource":{
                "executionPeriod":{
                    "end":"2021-08-14T00:07:48+00:00",
                    "start":"2021-08-13T00:07:48+00:00"
                },
                "identifier":[
                    {
                        "system":"http://bcda.cms.gov/api/v1/jobs",
                        "use":"official",
                        "value":"1"
                    }
                ],
                "input":[
                    {
                        "type":{
                            "text":"BULK FHIR Export"
                        },
                        "valueString":"GET https://bcda.test.gov/this-is-a-test"
                    }
                ],
                "intent":"order",
                "resourceType":"Task",
                "status":"completed"
            }
        }
    ],
    "resourceType":"Bundle",
    "total":1,
    "type":"searchset"
}
```

- [x] security checklist is completed for this change

[Security Checklist](https://confluence.cms.gov/display/BCDA/2021+-+08+-+18%3A+Jobs+Status+Endpoint+Security+Checklist)

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

The ACO can query the `GET .../jobs` endpoint(s) and retrieve historical data of all their jobs.
Postman tests added for CI and deployment.
Updated Swagger to include new endpoints under `jobs` section.

### Feedback Requested

General feedback as always.

### CODE REVIEW

File: bcda/api/requests_test.go [hunk 1]
```diff
@@ -303,3 +482,22 @@ func (s *RequestsTestSuite) genASRequest() *http.Request {
 
 	return req.WithContext(ctx)
 }
+
+func (s *RequestsTestSuite) genGetJobsRequest(statuses []models.JobStatus) *http.Request {
+	target := "http://bcda.cms.gov/api/v2/jobs"
```
Comment (katyasoup) [inline bcda/api/requests_test.go:487]: should this always be using v2?

**Review by katyasoup — COMMENTED** [review]
A couple questions that you can take or leave! Looks great!

File: bcda/responseutils/v2/writer.go [hunk 2]
```diff
@@ -15,6 +17,7 @@ import (
 	fhirmodelCR "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/resources/bundle_and_contained_resource_go_proto"
 	fhirmodelCS "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/resources/capability_statement_go_proto"
 	fhirmodelOO "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/resources/operation_outcome_go_proto"
+	fhirmodelT "github.com/google/fhir/go/proto/google/fhir/proto/r4/core/resources/task_go_proto"
```
[the remark this answers is not in the document]
Comment (SMLuthi) [reply bcda/responseutils/v2/writer.go:20]: HAHAHA

File: bcda/api/requests_test.go (hunk 1 above)
Comment (SMLuthi) [reply bcda/api/requests_test.go:501]: It doesnt *really* matter but I think its good to be consistent when it comes to the test. Will update.

**Review by katyasoup — APPROVED** [review]
🎉

**Review by jphechter — DISMISSED** [review]
This looks great!

File: bcda/api/alr.go [hunk 3]
```diff
@@ -31,7 +31,7 @@ func (h *Handler) alrRequest(w http.ResponseWriter, r *http.Request, reqType ser
 		panic("Request parameters must be set prior to calling this handler.")
 	}
 
-	if err := h.validateRequest(rp.ResourceTypes); err != nil {
+	if err := h.validateResources(rp.ResourceTypes); err != nil {
```
Comment (CosmicSyntax) [inline bcda/api/alr.go:34]: ❤️

File: bcda/api/requests.go [hunk 4]
```diff
@@ -155,6 +163,75 @@ func (h *Handler) buildBulkGroupRequest(r *http.Request) (service.RequestType, e
...
+		statusTypes = nil
+
+		// validate no duplicate status types
+		for _, status := range rawStatusTypes {
+			if _, ok := statusMap[status]; !ok {
+				statusMap[status] = struct{}{}
+				statusTypes = append(statusTypes, models.JobStatus(status))
+			} else {
+				errMsg := fmt.Sprintf("Repeated status type %s", status)
+				h.RespWriter.Exception(w, http.StatusBadRequest, responseutils.RequestErr, errMsg)
```
Comment (CosmicSyntax) [inline bcda/api/requests.go:187]: Perhaps we should look into making a ticket (for the H-team?) to see if it's better to dedup the list and just continue with the request?

File: bcda/api/requests_test.go [hunk 5]
```diff
@@ -125,19 +132,205 @@ func (s *RequestsTestSuite) TestRunoutEnabled() {
...
+						mockArgs = append(mockArgs, mock.Anything)
+					}
+				} else {
+					for k := range tt.statuses {
+						mockArgs = append(mockArgs, mock.Anything)
+						jobs = s.addNewJob(jobs, uint(k), tt.statuses[k], apiVersion)
+					}
+				}
+
+				mockSvc.On("GetJobs", mockArgs...).Return(
```
Comment (CosmicSyntax) [inline bcda/api/requests_test.go:184]: 🔥

File: bcda/api/v1/api_test.go [hunk 6]
```diff
@@ -441,6 +441,88 @@ func (s *APITestSuite) TestJobStatusWithWrongACO() {
...
+	req := httptest.NewRequest("GET", "/api/v1/jobs", nil)
+	ad := s.makeContextValues(acoUnderTest)
+	req = req.WithContext(context.WithValue(req.Context(), auth.AuthDataContextKey, ad))
+	rr := httptest.NewRecorder()
+
+	JobsStatus(rr, req)
+	assert.Equal(s.T(), http.StatusNotFound, rr.Code)
+}
+
+func (s *APITestSuite) TestJobsStatusNotFoundWithStatus() {
```
Comment (CosmicSyntax) [inline bcda/api/v1/api_test.go:472]: 🔥

File: bcda/responseutils/v2/writer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CosmicSyntax) [reply bcda/responseutils/v2/writer.go:20]: lol

**Review by CosmicSyntax — APPROVED** [review]
Amazing work as always!
🎉

**Review by katyasoup — APPROVED** [review]
🎉

**Pull request merged**